rfcs/000-consolidate-amundsen-microservices.md

```diff
@@ -0,0 +1,154 @@
+
+A few examples:
+
+| Package             | amundsenmetadata   | amundsenfrontend  |
+| ------------------  |:------------------:| -----------------:|
+| amundsen-common     | >=0.8.1            | ==0.6.0           |
+| flake8              | ==3.5.0            | ==3.8.4           |
+| flake8-tidy-imports | ==1.1.0            | ==4.2.1           |
+
+and many more... 
+
+*This change will put all the packages in just one place, making it easier to manage dependencies and no multiple variants of the same package.*
+
+#### 2: Development Efforts
+Having multiple repositories makes it really hard to implement a feature. Implementation and testing require efforts to synchronize and then code reviews, and finally, all the PRs across multiple repositories need to land in master at a certain time or at the same time. 
+
+*This change results in faster local development, one PR to fix a bug or a feature, no dependency hell as we are facing today, hence attract more contributors.*
+
+#### 3: Customization & Deployments
+One of the frequently mentioned topics re Amundsen is the complexity of the architecture, and the efforts required to customize Amundsen. Having multiple repositories results in multiple components to install, customize, manage and fix.
+
+*This change will make the customization much easier and the deployment will be pretty straightforward, with only 2 components. No multilevel docker files, easy maintenance of the infrastructure.* 
+
+Custom endpoints will still be injectable the same way we are doing today. We will write a blog/tutorial around each customization.
+
+#### 4: Code Duplication
+From requirements to config files, models, helper functions, exceptions handling, CI/CD pipelines, Docker files, and other repository management like licenses, PR/Issues templates, etc., is pretty much the same in all these microservices. Change in one repository deviates that repository from others at the moment. 
+
+*This change will move all the above into a single repository making it easy to change anything without keeping track of multiple repositories, and redundant code.* 
+
+Having one config file does NOT mean we can not have the incubator integration. It will be the same process as of today i.e., a separate incubation repo, which implements the required methods, and then install that as a requirement and set that as a proxy within the config file.
+
+
+## Transition Path
+Most of the code is already there and is spread across multiple repositories. I can see this happen in multiple phases.
+
+#### Phase I:
+Merge `amundsenmetadatalibrary` and `amundsensearchlibrary`. This should not take too long and will be easy to maintain the same URL patterns for each under separate directories within one repository. 
+The new repository will be named `amundsen`. During this period the `amundsenfrontend` repository will be unchanged and will work as it is. 
```

Comment:
There are multiple ways we can handle this. 
- Make use of the existing `amundsen` repository, and will start working in the same repo for the time being. We can have this in a separate directory, but this will break things for users once we move the code out of that directory to the root level. For now there is no setup.py or config or any such files that intersect with our implementation, therefore we can use the repo as is. This will certainly add the noise in the central repo.
- Have a `v2` branch for the time being, and have all the implementation in there, and after 2-3 months (this is the time we are expecting for this whole migration) replace the master branch with this v2 branch, and move today's master in a `legacy` branch.
- Make use of the `main` branch, and have all of our development in there. After we are done with the implementation, make `main` (which is not the default master on Github) the primary branch of the project.